Only in nw/v8: .git
diff -r -u --color up/v8/BUILD.gn nw/v8/BUILD.gn
--- up/v8/BUILD.gn	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/BUILD.gn	2026-04-25 12:56:39.498626656 +0000
@@ -62,7 +62,7 @@
   v8_deprecation_warnings = true
 
   # Enable compiler warnings when using V8_DEPRECATE_SOON apis.
-  v8_imminent_deprecation_warnings = true
+  v8_imminent_deprecation_warnings = false
 
   # Embeds the given script into the snapshot.
   v8_embed_script = ""
@@ -350,9 +350,9 @@
                                arm_control_flow_integrity != "none")
 
   # Enable heap reservation of size 4GB. Only possible for 64bit archs.
-  cppgc_enable_caged_heap =
-      v8_current_cpu == "x64" || v8_current_cpu == "arm64" ||
-      v8_current_cpu == "loong64" || v8_current_cpu == "riscv64"
+  cppgc_enable_caged_heap = false
+  #    v8_current_cpu == "x64" || v8_current_cpu == "arm64" ||
+  #    v8_current_cpu == "loong64" || v8_current_cpu == "riscv64"
 
   # Enables additional heap verification phases and checks.
   cppgc_enable_verify_heap = ""
@@ -362,7 +362,7 @@
 
   # Enable the V8 sandbox.
   # Sets -DV8_ENABLE_SANDBOX.
-  v8_enable_sandbox = ""
+  v8_enable_sandbox = false
 
   # Enable experimental hardware support for the V8 sandbox if available.
   # Sets -DV8_ENABLE_SANDBOX_HARDWARE_SUPPORT
@@ -401,7 +401,7 @@
   v8_enable_map_packing = false
 
   # Allow for JS promise hooks (instead of just C++).
-  v8_enable_javascript_promise_hooks = false
+  v8_enable_javascript_promise_hooks = true
 
   # Allow embedder data to be saved on continuations. Used to support
   # TaskAttribution and `scheduler.yield()`.
@@ -422,7 +422,7 @@
   # Enable legacy mode for ScriptOrModule's lifetime. By default it's a
   # temporary object, if enabled it will be kept alive by the parent Script.
   # This is only used by nodejs.
-  v8_scriptormodule_legacy_lifetime = false
+  v8_scriptormodule_legacy_lifetime = true
 
   # Enables pointer compression for 8GB heaps.
   # Sets -DV8_COMPRESS_POINTERS_8GB.
@@ -901,7 +901,7 @@
 
 config("internal_config") {
   defines = []
-
+  cflags = []
   # Only targets in this file and its subdirs can depend on this.
   visibility = [ "./*" ]
 
@@ -914,7 +914,8 @@
 
   if (is_component_build) {
     defines += [ "BUILDING_V8_SHARED_PRIVATE" ]
-  } else if (v8_expose_public_symbols) {
+  } #else if (v8_expose_public_symbols) {
+  if (true) {
     defines += [ "BUILDING_V8_SHARED" ]
   }
 
@@ -923,6 +924,10 @@
       libs = [ "atomic" ]
     }
   }
+
+  if (is_win) {
+    cflags += [ "/Zc:dllexportInlines-" ]
+  }
 }
 
 # Should be applied to all targets that write trace events.
@@ -939,14 +944,14 @@
 # This config should be applied to code using the libplatform.
 config("libplatform_config") {
   include_dirs = [ "include" ]
-  if (is_component_build) {
+  if (false) {
     defines = [ "USING_V8_PLATFORM_SHARED" ]
   }
 }
 
 # This config should be applied to code using the libbase.
 config("libbase_config") {
-  if (is_component_build) {
+  if (false) {
     defines = [ "USING_V8_BASE_SHARED" ]
   }
   libs = []
@@ -994,7 +999,10 @@
     defines += [
       "USING_V8_SHARED",
       "USING_V8_SHARED_PRIVATE",
+      "V8_SHARED", "USING_V8_PLATFORM_SHARED"
     ]
+  } else {
+    defines += [ "V8_SHARED" ]
   }
 
   if (current_cpu == "riscv64" || current_cpu == "riscv32") {
@@ -1195,7 +1203,7 @@
   # Pointer compression regresses binary size on Fuchsia by about 300K.
   # However, the change improves Oilpan memory by 15-20% (2-4% of PMF),
   # which is beneficial for memory-impoverished platforms.
-  cppgc_enable_pointer_compression = true
+  cppgc_enable_pointer_compression = false
 }
 if (cppgc_enable_young_generation) {
   enabled_external_cppgc_defines += [ "CPPGC_YOUNG_GENERATION" ]
@@ -3233,6 +3241,8 @@
   visibility = [
     ":*",
     "test/cctest:*",
+    "//tools/v8_context_snapshot:*",
+    "//chrome:*",
   ]
 
   allow_circular_includes_from = [ ":torque_generated_initializers" ]
@@ -7039,7 +7049,8 @@
   ]
 }
 
-v8_component("v8_libbase") {
+v8_static_lib("v8_libbase") {
+
   sources = [
     "src/base/abort-mode.cc",
     "src/base/abort-mode.h",
@@ -7231,7 +7242,7 @@
 
   defines = []
 
-  if (is_component_build) {
+  if (false) {
     defines = [ "BUILDING_V8_BASE_SHARED" ]
   }
 
@@ -7404,7 +7415,7 @@
   }
 }
 
-v8_component("v8_libplatform") {
+v8_static_lib("v8_libplatform") {
   sources = [
     "include/libplatform/libplatform-export.h",
     "include/libplatform/libplatform.h",
@@ -7439,7 +7450,7 @@
   v8_add_configs += [ "//build/config/compiler:thinlto_optimize_max" ]
   v8_remove_configs += [ "//build/config/compiler:thinlto_optimize_default" ]
 
-  if (is_component_build) {
+  if (true) {
     defines = [ "BUILDING_V8_PLATFORM_SHARED" ]
   }
 
@@ -7893,6 +7904,25 @@
 ###############################################################################
 # Executables
 #
+v8_executable("nwjc") {
+
+    sources = [
+      "src/nwjc.cc",
+    ]
+
+    configs = [
+      ":internal_config",
+    ]
+
+    deps = [
+      ":v8_base",
+      ":v8_libplatform",
+      ":v8_libbase",
+      ":v8_snapshot",
+      ":v8_initializers",
+      "//build/win:default_exe_manifest",
+    ]
+}
 
 if (current_toolchain == v8_generator_toolchain) {
   v8_executable("bytecode_builtins_list_generator") {
diff -r -u --color up/v8/gni/v8.gni nw/v8/gni/v8.gni
--- up/v8/gni/v8.gni	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/gni/v8.gni	2026-04-25 12:56:39.501422635 +0000
@@ -477,6 +477,15 @@
   }
 }
 
+template("v8_static_lib") {
+  static_library(target_name) {
+    forward_variables_from(invoker, "*", [ "configs" ])
+    configs += invoker.configs
+    configs -= v8_remove_configs
+    configs += v8_add_configs
+  }
+}
+
 template("v8_executable") {
   if (is_ios) {
     import("//build/config/ios/rules.gni")
Only in nw/v8: gypfiles
diff -r -u --color up/v8/include/libplatform/v8-tracing.h nw/v8/include/libplatform/v8-tracing.h
--- up/v8/include/libplatform/v8-tracing.h	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/include/libplatform/v8-tracing.h	2026-04-25 12:56:39.504522791 +0000
@@ -35,7 +35,7 @@
 
 const int kTraceMaxNumArgs = 2;
 
-class V8_PLATFORM_EXPORT TraceObject {
+class V8_EXPORT TraceObject {
  public:
   union ArgValue {
     uint64_t as_uint;
@@ -46,7 +46,7 @@
   };
 
   TraceObject() = default;
-  ~TraceObject();
+  ~TraceObject() { delete[] parameter_copy_storage_; }
   void Initialize(
       char phase, const uint8_t* category_enabled_flag, const char* name,
       const char* scope, uint64_t id, uint64_t bind_id, int num_args,
diff -r -u --color up/v8/include/v8-array-buffer.h nw/v8/include/v8-array-buffer.h
--- up/v8/include/v8-array-buffer.h	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/include/v8-array-buffer.h	2026-04-25 12:56:39.504522791 +0000
@@ -186,7 +186,8 @@
      * while kReservation is for larger allocations with the ability to set
      * access permissions.
      */
-    enum class AllocationMode { kNormal, kReservation };
+    enum class AllocationMode { kNormal, kReservation, kNodeJS };
+    virtual void Free(void* data, size_t length, AllocationMode mode);
 
     /**
      * Returns page allocator used by this Allocator instance.
@@ -273,6 +274,8 @@
    */
   static Local<ArrayBuffer> New(Isolate* isolate,
                                 std::shared_ptr<BackingStore> backing_store);
+  static Local<ArrayBuffer> NewNode(Isolate* isolate,
+                                std::shared_ptr<BackingStore> backing_store);
 
   /**
    * Returns a new standalone BackingStore that is allocated using the array
@@ -349,6 +352,8 @@
       "ok).")
   void Detach();
 
+  void set_nodejs(bool);
+
   /**
    * Detaches this ArrayBuffer and all its views (typed arrays).
    * Detaching sets the byte length of the buffer and all typed arrays to zero,
diff -r -u --color up/v8/include/v8-context.h nw/v8/include/v8-context.h
--- up/v8/include/v8-context.h	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/include/v8-context.h	2026-04-25 12:56:39.504522791 +0000
@@ -255,6 +255,12 @@
    */
   Maybe<void> DeepFreeze(DeepFreezeDelegate* delegate = nullptr);
 
+  /** Returns the isolate associated with a current context. */
+  V8_DEPRECATED(
+      "Use Isolate::GetCurrent() instead, which is guaranteed to return the "
+      "same isolate since https://crrev.com/c/6458560.")
+  Isolate* GetIsolate();
+
   /** Returns the microtask queue associated with a current context. */
   MicrotaskQueue* GetMicrotaskQueue();
 
diff -r -u --color up/v8/include/v8-initialization.h nw/v8/include/v8-initialization.h
--- up/v8/include/v8-initialization.h	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/include/v8-initialization.h	2026-04-25 12:56:39.504626601 +0000
@@ -174,6 +174,7 @@
    */
   static bool InitializeICUDefaultLocation(const char* exec_path,
                                            const char* icu_data_file = nullptr);
+  static void* RawICUData();
 
   /**
    * Initialize the external startup data. The embedder only needs to
diff -r -u --color up/v8/include/v8-internal.h nw/v8/include/v8-internal.h
--- up/v8/include/v8-internal.h	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/include/v8-internal.h	2026-04-25 12:56:39.504626601 +0000
@@ -1464,6 +1464,18 @@
 #endif
   }
 
+  V8_DEPRECATED(
+      "Use GetCurrentIsolateForSandbox() instead, which is guaranteed to "
+      "return the same isolate since https://crrev.com/c/6458560.")
+  V8_INLINE static v8::Isolate* GetIsolateForSandbox(Address obj) {
+#ifdef V8_ENABLE_SANDBOX
+    return GetCurrentIsolate();
+#else
+    // Not used in non-sandbox mode.
+    return nullptr;
+#endif
+  }
+
   // Returns v8::Isolate::Current(), but without needing to include the
   // v8-isolate.h header.
   V8_EXPORT static v8::Isolate* GetCurrentIsolate();
diff -r -u --color up/v8/include/v8-isolate.h nw/v8/include/v8-isolate.h
--- up/v8/include/v8-isolate.h	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/include/v8-isolate.h	2026-04-25 12:56:39.505626591 +0000
@@ -290,6 +290,7 @@
  */
 class V8_EXPORT Isolate {
  public:
+  ArrayBuffer::Allocator* array_buffer_allocator();
   /**
    * Initial configuration parameters for a new Isolate.
    */
@@ -370,6 +371,7 @@
     CppHeap* cpp_heap = nullptr;
   };
 
+  void SetArrayBufferAllocatorShared(std::shared_ptr<ArrayBuffer::Allocator> allocator);
   /**
    * Stack-allocated class which sets the isolate for all operations
    * executed within a local scope.
@@ -1238,6 +1240,28 @@
    */
   void SetEmbedderRootsHandler(EmbedderRootsHandler* handler);
 
+  /**
+   * Attaches a managed C++ heap as an extension to the JavaScript heap. The
+   * embedder maintains ownership of the CppHeap. At most one C++ heap can be
+   * attached to V8.
+   *
+   * Multi-threaded use requires the use of v8::Locker/v8::Unlocker, see
+   * CppHeap.
+   *
+   * If a CppHeap is set via CreateParams, then this call is a noop.
+   */
+  V8_DEPRECATED("Set the heap on Isolate creation using CreateParams instead.")
+  void AttachCppHeap(CppHeap*);
+
+  /**
+   * Detaches a managed C++ heap if one was attached using `AttachCppHeap()`.
+   *
+   * If a CppHeap is set via CreateParams, then this call is a noop.
+   */
+  V8_DEPRECATED(
+      "The CppHeap gets detached automatically during Isolate tear down.")
+  void DetachCppHeap();
+
   using ReleaseCppHeapCallback = void (*)(std::unique_ptr<CppHeap>);
 
   /**
diff -r -u --color up/v8/include/v8-message.h nw/v8/include/v8-message.h
--- up/v8/include/v8-message.h	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/include/v8-message.h	2026-04-25 12:56:39.505626591 +0000
@@ -108,6 +108,14 @@
  public:
   Local<String> Get() const;
 
+  /**
+   * Return the isolate to which the Message belongs.
+   */
+  V8_DEPRECATED(
+      "Use Isolate::GetCurrent() instead, which is guaranteed to return the "
+      "same isolate since https://crrev.com/c/6458560.")
+  Isolate* GetIsolate() const;
+
   V8_WARN_UNUSED_RESULT MaybeLocal<String> GetSource(
       Local<Context> context) const;
   V8_WARN_UNUSED_RESULT MaybeLocal<String> GetSourceLine(
diff -r -u --color up/v8/include/v8-object.h nw/v8/include/v8-object.h
--- up/v8/include/v8-object.h	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/include/v8-object.h	2026-04-25 12:56:39.505626591 +0000
@@ -871,6 +871,21 @@
       Local<Context> context, int argc, Local<Value> argv[]);
 
   /**
+   * Return the isolate to which the Object belongs to.
+   */
+  V8_DEPRECATED(
+      "Use Isolate::GetCurrent() instead, which is guaranteed to return the "
+      "same isolate since https://crrev.com/c/6458560.")
+  Isolate* GetIsolate();
+
+  V8_DEPRECATED(
+      "Use Isolate::GetCurrent() instead, which is guaranteed to return the "
+      "same isolate since https://crrev.com/c/6458560.")
+  V8_INLINE static Isolate* GetIsolate(const TracedReference<Object>& handle) {
+    return handle.template value<Object>()->GetIsolate();
+  }
+
+  /**
    * If this object is a Set, Map, WeakSet or WeakMap, this returns a
    * representation of the elements of this object as an array.
    * If this object is a SetIterator or MapIterator, this returns all
diff -r -u --color up/v8/include/v8-platform.h nw/v8/include/v8-platform.h
--- up/v8/include/v8-platform.h	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/include/v8-platform.h	2026-04-25 12:56:39.506626582 +0000
@@ -1546,6 +1546,7 @@
       TaskPriority priority, std::unique_ptr<Task> task,
       double delay_in_seconds, const SourceLocation& location) = 0;
 };
+void V8_EXPORT SetTLSPlatform(Platform* platform);
 
 }  // namespace v8
 
diff -r -u --color up/v8/include/v8-script.h nw/v8/include/v8-script.h
--- up/v8/include/v8-script.h	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/include/v8-script.h	2026-04-25 12:56:39.506626582 +0000
@@ -816,6 +816,7 @@
       CompileOptions options = kNoCompileOptions,
       NoCacheReason no_cache_reason = kNoCacheNoReason);
 
+  static MaybeLocal<Module> CompileModuleWithCache(Isolate* isolate, Source* source);
   /**
    * Returns a task which streams script data into V8, or NULL if the script
    * cannot be streamed. The user is responsible for running the task on a
@@ -1006,6 +1007,9 @@
   return reinterpret_cast<Module*>(data);
 }
 
+void V8_EXPORT FixSourceNWBin(Isolate* v8_isolate, Local<UnboundScript> script);
+void V8_EXPORT FixSourceNWBin(Isolate* v8_isolate, Local<Module> module);
+
 }  // namespace v8
 
 #endif  // INCLUDE_V8_SCRIPT_H_
diff -r -u --color up/v8/include/v8-wasm.h nw/v8/include/v8-wasm.h
--- up/v8/include/v8-wasm.h	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/include/v8-wasm.h	2026-04-25 12:56:39.507626573 +0000
@@ -123,6 +123,9 @@
   static void CheckCast(Value* obj);
 };
 
+V8_DEPRECATED("Use WasmModuleObject")
+              typedef WasmModuleObject WasmCompiledModule;
+
 /**
  * The V8 interface for WebAssembly streaming compilation. When streaming
  * compilation is initiated, V8 passes a {WasmStreaming} object to the embedder
diff -r -u --color up/v8/include/v8config.h nw/v8/include/v8config.h
--- up/v8/include/v8config.h	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/include/v8config.h	2026-04-25 12:56:39.507626573 +0000
@@ -396,7 +396,7 @@
          && !defined(_WIN32)))                         /* not on windows */    \
      && !defined(COMPONENT_BUILD)                      /* no component build */\
      && __clang_major__ >= 17                          /* clang >= 17 */
-# define V8_HAS_ATTRIBUTE_PRESERVE_MOST (__has_attribute(preserve_most))
+//# define V8_HAS_ATTRIBUTE_PRESERVE_MOST (__has_attribute(preserve_most))
 #endif
 # define V8_HAS_ATTRIBUTE_VISIBILITY (__has_attribute(visibility))
 # define V8_HAS_ATTRIBUTE_WARN_UNUSED_RESULT \
@@ -820,8 +820,10 @@
 
 // Exposing private symbols requires exposing public symbols too.
 #ifdef BUILDING_V8_SHARED_PRIVATE
+#ifndef BUILDING_V8_SHARED
 #define BUILDING_V8_SHARED
 #endif
+#endif
 
 #if defined(BUILDING_V8_SHARED) && defined(USING_V8_SHARED)
 #error Inconsistent build configuration: To build the V8 shared library \
diff -r -u --color up/v8/src/api/api.cc nw/v8/src/api/api.cc
--- up/v8/src/api/api.cc	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/src/api/api.cc	2026-04-25 12:56:39.510626546 +0000
@@ -325,6 +325,11 @@
   i::V8::SetSnapshotBlob(snapshot_blob);
 }
 
+void v8::ArrayBuffer::Allocator::Free(void* data, size_t length,
+                                      AllocationMode mode) {
+  UNIMPLEMENTED();
+}
+
 namespace {
 
 #ifdef V8_ENABLE_SANDBOX
@@ -2686,6 +2691,21 @@
   return ToApiHandle<Module>(i_isolate->factory()->NewSourceTextModule(shared));
 }
 
+MaybeLocal<Module> ScriptCompiler::CompileModuleWithCache(Isolate* isolate,
+                                                          Source* source) {
+  i::Isolate* i_isolate = reinterpret_cast<i::Isolate*>(isolate);
+
+  Utils::ApiCheck(source->GetResourceOptions().IsModule(),
+                  "v8::ScriptCompiler::CompileModule",
+                  "Invalid ScriptOrigin: is_module must be true");
+  auto maybe = CompileUnboundInternal(isolate, source, kConsumeCodeCache, kNoCacheNoReason);
+  Local<UnboundScript> unbound;
+  if (!maybe.ToLocal(&unbound)) return MaybeLocal<Module>();
+
+  i::Handle<i::SharedFunctionInfo> shared = Utils::OpenHandle(*unbound);
+  return ToApiHandle<Module>(i_isolate->factory()->NewSourceTextModule(shared));
+}
+
 // static
 V8_WARN_UNUSED_RESULT MaybeLocal<Function> ScriptCompiler::CompileFunction(
     Local<Context> v8_context, Source* source, size_t arguments_count,
@@ -3122,6 +3142,11 @@
   return scope.Escape(result);
 }
 
+v8::Isolate* Message::GetIsolate() const {
+  i::Isolate* i_isolate = i::Isolate::Current();
+  return reinterpret_cast<Isolate*>(i_isolate);
+}
+
 ScriptOrigin Message::GetScriptOrigin() const {
   auto self = Utils::OpenDirectHandle(this);
   i::Isolate* i_isolate = i::Isolate::Current();
@@ -3361,8 +3386,8 @@
   i::Isolate* i_isolate = i::Isolate::Current();
   i::DirectHandle<i::Script> script(self->script(), i_isolate);
   i::Script::PositionInfo info;
-  CHECK(i::Script::GetPositionInfo(
-      script, i::StackFrameInfo::GetSourcePosition(self), &info));
+  i::Script::GetPositionInfo(
+     script, i::StackFrameInfo::GetSourcePosition(self), &info);
   if (script->HasSourceURLComment()) {
     info.line -= script->line_offset();
     if (info.line == 0) {
@@ -6500,6 +6525,10 @@
   return i::InitializeICUDefaultLocation(exec_path, icu_data_file);
 }
 
+void* v8::V8::RawICUData() {
+  return i::RawICUData();
+}
+
 void v8::V8::InitializeExternalStartupData(const char* directory_path) {
   i::InitializeExternalStartupData(directory_path);
 }
@@ -7139,6 +7168,8 @@
   return JustVoid();
 }
 
+v8::Isolate* Context::GetIsolate() { return Isolate::GetCurrent(); }
+
 v8::MicrotaskQueue* Context::GetMicrotaskQueue() {
   auto env = Utils::OpenDirectHandle(this);
   Utils::ApiCheck(i::IsNativeContext(*env), "v8::Context::GetMicrotaskQueue",
@@ -7746,6 +7777,11 @@
   return self->Equals(*other);
 }
 
+Isolate* v8::Object::GetIsolate() {
+  i::Isolate* i_isolate = i::Isolate::Current();
+  return reinterpret_cast<Isolate*>(i_isolate);
+}
+
 Local<v8::Object> v8::Object::New(Isolate* v8_isolate) {
   i::Isolate* i_isolate = reinterpret_cast<i::Isolate*>(v8_isolate);
   ApiRuntimeCallStatsScope rcs_scope(i_isolate, RCCId::kAPI_Object_New);
@@ -9212,6 +9248,25 @@
       static_cast<v8::BackingStore*>(backing_store.release()));
 }
 
+Local<ArrayBuffer> v8::ArrayBuffer::NewNode(
+    Isolate* isolate, std::shared_ptr<BackingStore> backing_store) {
+  CHECK_IMPLIES(backing_store->ByteLength() != 0,
+                backing_store->Data() != nullptr);
+  i::Isolate* i_isolate = reinterpret_cast<i::Isolate*>(isolate);
+  ApiRuntimeCallStatsScope rcs_scope(i_isolate, RCCId::kAPI_ArrayBuffer_New);
+  EnterV8NoScriptNoExceptionScope api_scope(i_isolate);
+  std::shared_ptr<i::BackingStore> i_backing_store(
+      ToInternal(std::move(backing_store)));
+  Utils::ApiCheck(
+      !i_backing_store->is_shared(), "v8_ArrayBuffer_New",
+      "Cannot construct ArrayBuffer with a BackingStore of SharedArrayBuffer");
+  i_backing_store->set_nodejs(true);
+  i::Handle<i::JSArrayBuffer> obj =
+      i_isolate->factory()->NewJSArrayBuffer(std::move(i_backing_store));
+  obj->set_is_node_js(true);
+  return Utils::ToLocal(obj);
+}
+
 Local<ArrayBuffer> v8::ArrayBufferView::Buffer() {
   auto obj = Utils::OpenDirectHandle(this);
   i::Isolate* i_isolate = i::Isolate::Current();
@@ -9827,6 +9882,11 @@
   i_isolate->ClearKeptObjects();
 }
 
+ArrayBuffer::Allocator* Isolate::array_buffer_allocator() {
+  i::Isolate* isolate = reinterpret_cast<i::Isolate*>(this);
+  return isolate->array_buffer_allocator();
+}
+
 v8::Local<v8::Context> Isolate::GetCurrentContext() {
   i::Isolate* i_isolate = reinterpret_cast<i::Isolate*>(this);
   i::Tagged<i::Context> context = i_isolate->context();
@@ -9940,6 +10000,16 @@
   i_isolate->heap()->SetEmbedderRootsHandler(handler);
 }
 
+void Isolate::AttachCppHeap(CppHeap* cpp_heap) {
+  i::Isolate* i_isolate = reinterpret_cast<i::Isolate*>(this);
+  i_isolate->heap()->AttachCppHeap(cpp_heap);
+}
+
+void Isolate::DetachCppHeap() {
+  i::Isolate* i_isolate = reinterpret_cast<i::Isolate*>(this);
+  i_isolate->heap()->DetachCppHeap();
+}
+
 CppHeap* Isolate::GetCppHeap() const {
   const i::Isolate* i_isolate = reinterpret_cast<const i::Isolate*>(this);
   return i_isolate->heap()->cpp_heap();
@@ -10052,6 +10122,13 @@
 
 Isolate::CreateParams::~CreateParams() = default;
 
+void Isolate::SetArrayBufferAllocatorShared(
+                                            std::shared_ptr<ArrayBuffer::Allocator> allocator) {
+  i::Isolate* isolate = reinterpret_cast<i::Isolate*>(this);
+  CHECK_EQ(allocator.get(), isolate->array_buffer_allocator());
+  isolate->set_array_buffer_allocator_shared(std::move(allocator));
+}
+
 // static
 // This is separate so that tests can provide a different |isolate|.
 void Isolate::Initialize(Isolate* v8_isolate,
@@ -12134,6 +12211,29 @@
 }
 #endif  // !V8_ENABLE_WEBASSEMBLY
 
+void SetTLSPlatform(Platform* platform) {
+  i::V8::SetTLSPlatform(platform);
+}
+
+void FixSourceNWBin(Isolate* v8_isolate, Local<UnboundScript> script) {
+  i::Isolate* isolate = reinterpret_cast<i::Isolate*>(v8_isolate);
+  i::Handle<i::HeapObject> obj =
+    i::Cast<i::HeapObject>(v8::Utils::OpenHandle(*script));
+  i::Handle<i::SharedFunctionInfo>
+    function_info(i::Cast<i::SharedFunctionInfo>(*obj), isolate);
+  i::Handle<i::Script> iscript(i::Cast<i::Script>(function_info->script()),
+                              isolate);
+  iscript->set_source(i::ReadOnlyRoots(isolate).undefined_value());
+}
+
+void FixSourceNWBin(Isolate* v8_isolate, Local<Module> module) {
+  i::Isolate* isolate = reinterpret_cast<i::Isolate*>(v8_isolate);
+  i::Handle<i::SourceTextModule> obj =
+    i::Cast<i::SourceTextModule>(v8::Utils::OpenHandle(*module));
+  i::Handle<i::Script> iscript(i::Cast<i::Script>(obj->GetScript()), isolate);
+  iscript->set_source(i::ReadOnlyRoots(isolate).undefined_value());
+}
+
 namespace internal {
 
 const size_t HandleScopeImplementer::kEnteredContextsOffset =
diff -r -u --color up/v8/src/codegen/compiler.cc nw/v8/src/codegen/compiler.cc
--- up/v8/src/codegen/compiler.cc	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/src/codegen/compiler.cc	2026-04-25 12:56:39.556626124 +0000
@@ -4001,7 +4001,9 @@
     maybe_script = lookup_result.script();
     maybe_result = lookup_result.toplevel_sfi();
     is_compiled_scope = lookup_result.is_compiled_scope();
-    if (!maybe_result.is_null()) {
+    //NWJS#5168: will hit previous cache, use 0 source_length trick to
+    //bypass and try to consume cache
+    if (!maybe_result.is_null() && source->length()) {
       compile_timer.set_hit_isolate_cache();
     } else if (can_consume_code_cache) {
       compile_timer.set_consuming_code_cache();
diff -r -u --color up/v8/src/common/thread-local-storage.h nw/v8/src/common/thread-local-storage.h
--- up/v8/src/common/thread-local-storage.h	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/src/common/thread-local-storage.h	2026-04-25 12:56:39.576625941 +0000
@@ -7,7 +7,7 @@
 
 #include "include/v8config.h"
 
-#if defined(COMPONENT_BUILD) || defined(V8_TLS_USED_IN_LIBRARY)
+#if 1 //defined(COMPONENT_BUILD) || defined(V8_TLS_USED_IN_LIBRARY)
 #define V8_TLS_LIBRARY_MODE 1
 #else
 #define V8_TLS_LIBRARY_MODE 0
diff -r -u --color up/v8/src/debug/debug-scopes.cc nw/v8/src/debug/debug-scopes.cc
--- up/v8/src/debug/debug-scopes.cc	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/src/debug/debug-scopes.cc	2026-04-25 12:56:39.628771646 +0000
@@ -313,7 +313,10 @@
   info_ = std::make_unique<ParseInfo>(isolate_, flags, &compile_state,
                                       reusable_compile_state_.get());
 
-  const bool parse_result =
+  bool has_source_code = shared_info->HasSourceCode();
+  bool parse_result = false;
+  if (has_source_code)
+    parse_result =
       flags.is_toplevel()
           ? parsing::ParseProgram(info_.get(), script, maybe_outer_scope,
                                   isolate_, parsing::ReportStatisticsMode::kNo)
diff -r -u --color up/v8/src/flags/flag-definitions.h nw/v8/src/flags/flag-definitions.h
--- up/v8/src/flags/flag-definitions.h	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/src/flags/flag-definitions.h	2026-04-25 12:56:39.650670356 +0000
@@ -3224,7 +3224,7 @@
 DEFINE_IMPLICATION(log_maps, log_code)
 
 // parser.cc
-DEFINE_BOOL(allow_natives_syntax, false, "allow natives syntax")
+DEFINE_BOOL(allow_natives_syntax, true, "allow natives syntax")
 DEFINE_BOOL(allow_natives_for_differential_fuzzing, false,
             "allow only natives explicitly allowlisted for differential "
             "fuzzers")
@@ -3727,6 +3727,8 @@
             "Print JS and Wasm feature flags grouped by in-progress, staged, "
             "and shipped")
 
+DEFINE_BOOL(nw_module, false, "Whether the input file is a module")
+
 // Slow histograms are also enabled via --dump-counters in d8.
 DEFINE_BOOL(slow_histograms, false,
             "Enable slow histograms with more overhead.")
diff -r -u --color up/v8/src/heap/cppgc-js/cpp-heap.cc nw/v8/src/heap/cppgc-js/cpp-heap.cc
--- up/v8/src/heap/cppgc-js/cpp-heap.cc	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/src/heap/cppgc-js/cpp-heap.cc	2026-04-25 12:56:39.654625227 +0000
@@ -537,6 +537,11 @@
 }
 
 CppHeap::~CppHeap() {
+  if (isolate_) {
+    // TODO(ahaas): Delete this code once `v8::Isolate::DetachCppHeap` has been
+    // deleted.
+    isolate_->heap()->DetachCppHeap();
+  }
   Terminate();
 }
 
diff -r -u --color up/v8/src/heap/heap.cc nw/v8/src/heap/heap.cc
--- up/v8/src/heap/heap.cc	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/src/heap/heap.cc	2026-04-25 12:56:39.662625154 +0000
@@ -6332,6 +6332,21 @@
   cpp_heap_ = cpp_heap;
 }
 
+void Heap::DetachCppHeap() {
+  // The API function should be a noop in case a CppHeap was passed on Isolate
+  // creation.
+  if (owning_cpp_heap_) {
+    return;
+  }
+
+  // The CppHeap may have been detached already.
+  if (!cpp_heap_) return;
+
+  CppHeap::From(cpp_heap_)->StartDetachingIsolate();
+  CppHeap::From(cpp_heap_)->DetachIsolate();
+  cpp_heap_ = nullptr;
+}
+
 std::optional<StackState> Heap::overridden_stack_state() const {
   if (!embedder_stack_state_origin_) return {};
   return embedder_stack_state_;
diff -r -u --color up/v8/src/heap/heap.h nw/v8/src/heap/heap.h
--- up/v8/src/heap/heap.h	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/src/heap/heap.h	2026-04-25 12:56:39.663625144 +0000
@@ -1151,6 +1151,9 @@
   // Unified heap (C++) support. ===============================================
   // ===========================================================================
 
+  V8_EXPORT_PRIVATE void AttachCppHeap(v8::CppHeap* cpp_heap);
+  V8_EXPORT_PRIVATE void DetachCppHeap();
+
   v8::CppHeap* cpp_heap() const { return cpp_heap_; }
 
   std::optional<StackState> overridden_stack_state() const;
@@ -1716,8 +1719,6 @@
  private:
   class AllocationTrackerForDebugging;
 
-  void AttachCppHeap(v8::CppHeap* cpp_heap);
-
   using ExternalStringTableUpdaterCallback =
       Tagged<String> (*)(Heap* heap, FullObjectSlot pointer);
 
diff -r -u --color up/v8/src/heap/marking-visitor-inl.h nw/v8/src/heap/marking-visitor-inl.h
--- up/v8/src/heap/marking-visitor-inl.h	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/src/heap/marking-visitor-inl.h	2026-04-25 12:56:39.666625117 +0000
@@ -527,6 +527,13 @@
     return false;
   }
 
+  Tagged<Object> script_obj = sfi->script();
+  if (!i::IsUndefined(script_obj)) {
+    Tagged<Script> script = Cast<Script>(script_obj);
+    if (i::IsUndefined(script->source()))
+      return false;
+  }
+
   return IsBytecodeArray(data);
 }
 
diff -r -u --color up/v8/src/init/icu_util.cc nw/v8/src/init/icu_util.cc
--- up/v8/src/init/icu_util.cc	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/src/init/icu_util.cc	2026-04-25 12:56:39.674788149 +0000
@@ -36,6 +36,10 @@
 }  // namespace
 #endif
 
+void* RawICUData() {
+  return (void*)g_icu_data_ptr;
+}
+
 bool InitializeICUDefaultLocation(const char* exec_path,
                                   const char* icu_data_file) {
 #if !defined(V8_INTL_SUPPORT)
diff -r -u --color up/v8/src/init/icu_util.h nw/v8/src/init/icu_util.h
--- up/v8/src/init/icu_util.h	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/src/init/icu_util.h	2026-04-25 12:56:39.674788149 +0000
@@ -20,6 +20,8 @@
 V8_WARN_UNUSED_RESULT bool InitializeICUDefaultLocation(
     const char* exec_path, const char* icu_data_file);
 
+void* RawICUData();
+
 }  // namespace internal
 }  // namespace v8
 
diff -r -u --color up/v8/src/init/startup-data-util.cc nw/v8/src/init/startup-data-util.cc
--- up/v8/src/init/startup-data-util.cc	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/src/init/startup-data-util.cc	2026-04-25 12:56:39.675625034 +0000
@@ -77,10 +77,23 @@
 
 void InitializeExternalStartupData(const char* directory_path) {
 #ifdef V8_USE_EXTERNAL_STARTUP_DATA
+#if 0
   const char* snapshot_name = "snapshot_blob.bin";
   std::unique_ptr<char[]> snapshot =
       base::RelativePath(directory_path, snapshot_name);
   LoadFromFile(snapshot.get());
+#endif
+#ifdef __APPLE__
+#if V8_TARGET_ARCH_X64
+  const char* snapshot_name = "v8_context_snapshot.x86_64.bin";
+#else
+  const char* snapshot_name = "v8_context_snapshot.arm64.bin";
+#endif
+#else
+  const char* snapshot_name = "v8_context_snapshot.bin";
+#endif
+  std::unique_ptr<char[]> snapshot = base::RelativePath(directory_path, snapshot_name);
+  LoadFromFile(snapshot.get());
 #endif  // V8_USE_EXTERNAL_STARTUP_DATA
 }
 
diff -r -u --color up/v8/src/init/v8.cc nw/v8/src/init/v8.cc
--- up/v8/src/init/v8.cc	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/src/init/v8.cc	2026-04-25 12:56:39.675625034 +0000
@@ -105,11 +105,20 @@
 #endif
 
 // static
+base::Thread::LocalStorageKey platform_tls_key_;
+void V8::SetTLSPlatform(v8::Platform* platform) {
+  base::Thread::SetThreadLocal(platform_tls_key_, platform);
+}
+
 void V8::InitializePlatform(v8::Platform* platform) {
   AdvanceStartupState(V8StartupState::kPlatformInitializing);
   CHECK(!platform_);
   CHECK_NOT_NULL(platform);
   platform_ = platform;
+
+  platform_tls_key_ = base::Thread::CreateThreadLocalKey();
+  base::Thread::SetThreadLocal(platform_tls_key_, platform);
+
   v8::base::SetPrintStackTrace(platform_->GetStackTracePrinter());
 #if defined(V8_USE_PERFETTO)
   // TrackEvent must be registered before TracingCategoryObserver::SetUp().
@@ -248,7 +257,6 @@
 #endif
   IsolateGroup::InitializeOncePerProcess();
   Isolate::InitializeOncePerProcess();
-
 #if defined(USE_SIMULATOR)
   Simulator::InitializeOncePerProcess();
 #endif
@@ -310,7 +318,11 @@
 }
 
 v8::Platform* V8::GetCurrentPlatform() {
-  v8::Platform* platform = reinterpret_cast<v8::Platform*>(
+  v8::Platform* platform;
+  platform = reinterpret_cast<v8::Platform*>(base::Thread::GetThreadLocal(platform_tls_key_));
+  if (platform)
+    return platform;
+  platform = reinterpret_cast<v8::Platform*>(
       base::Relaxed_Load(reinterpret_cast<base::AtomicWord*>(&platform_)));
   DCHECK(platform);
   return platform;
diff -r -u --color up/v8/src/init/v8.h nw/v8/src/init/v8.h
--- up/v8/src/init/v8.h	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/src/init/v8.h	2026-04-25 12:56:39.675625034 +0000
@@ -36,6 +36,8 @@
   V8_EXPORT_PRIVATE static const OOMDetails kNoOOMDetails;
   V8_EXPORT_PRIVATE static const OOMDetails kHeapOOM;
 
+  static void SetTLSPlatform(v8::Platform* platform);
+
   static void InitializePlatform(v8::Platform* platform);
   V8_EXPORT_PRIVATE static void InitializePlatformForTesting(
       v8::Platform* platform);
diff -r -u --color up/v8/src/inspector/v8-heap-profiler-agent-impl.cc nw/v8/src/inspector/v8-heap-profiler-agent-impl.cc
--- up/v8/src/inspector/v8-heap-profiler-agent-impl.cc	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/src/inspector/v8-heap-profiler-agent-impl.cc	2026-04-25 12:56:39.677625016 +0000
@@ -389,6 +389,15 @@
     progress.reset(new HeapSnapshotProgress(&m_frontend));
 
   ContextNameResolver resolver(m_session);
+#ifdef __APPLE__
+    // exit the context we entered in g_uv_runloop_once or taking
+    // snapshot will fail.
+    v8::Isolate* isolate = v8::Isolate::GetCurrent();
+    v8::HandleScope handle_scope(isolate);
+    v8::Local<v8::Context> context = isolate->GetEnteredOrMicrotaskContext();
+    if (!context.IsEmpty())
+      context->Exit();
+#endif
   v8::HeapProfiler::HeapSnapshotOptions options;
   options.context_name_resolver = &resolver;
   options.control = progress.get();
@@ -405,6 +414,10 @@
           : v8::HeapProfiler::NumericsMode::kHideNumericValues;
   options.stack_state = stackState;
   const v8::HeapSnapshot* snapshot = profiler->TakeHeapSnapshot(options);
+#ifdef __APPLE__
+    if (!context.IsEmpty())
+      context->Enter();
+#endif
   if (!snapshot) return Response::ServerError("Failed to take heap snapshot");
   HeapSnapshotOutputStream stream(&m_frontend);
   snapshot->Serialize(&stream);
diff -r -u --color up/v8/src/inspector/v8-stack-trace-impl.cc nw/v8/src/inspector/v8-stack-trace-impl.cc
--- up/v8/src/inspector/v8-stack-trace-impl.cc	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/src/inspector/v8-stack-trace-impl.cc	2026-04-25 12:56:39.678625007 +0000
@@ -164,8 +164,8 @@
       m_lineNumber(lineNumber),
       m_columnNumber(columnNumber),
       m_hasSourceURLComment(hasSourceURLComment) {
-  DCHECK_NE(v8::Message::kNoLineNumberInfo, m_lineNumber + 1);
-  DCHECK_NE(v8::Message::kNoColumnInfo, m_columnNumber + 1);
+  //DCHECK_NE(v8::Message::kNoLineNumberInfo, m_lineNumber + 1);
+  //DCHECK_NE(v8::Message::kNoColumnInfo, m_columnNumber + 1);
 }
 
 const String16& StackFrame::functionName() const { return m_functionName; }
diff -r -u --color up/v8/src/libplatform/tracing/trace-object.cc nw/v8/src/libplatform/tracing/trace-object.cc
--- up/v8/src/libplatform/tracing/trace-object.cc	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/src/libplatform/tracing/trace-object.cc	2026-04-25 12:56:39.685308949 +0000
@@ -99,7 +99,7 @@
   }
 }
 
-TraceObject::~TraceObject() { delete[] parameter_copy_storage_; }
+//TraceObject::~TraceObject() { delete[] parameter_copy_storage_; }
 
 void TraceObject::UpdateDuration(int64_t timestamp, int64_t cpu_timestamp) {
   duration_ = timestamp - ts_;
Only in nw/v8/src: nwjc.cc
diff -r -u --color up/v8/src/objects/backing-store.cc nw/v8/src/objects/backing-store.cc
--- up/v8/src/objects/backing-store.cc	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/src/objects/backing-store.cc	2026-04-25 12:56:39.700044916 +0000
@@ -112,7 +112,8 @@
       byte_length_(byte_length),
       max_byte_length_(max_byte_length),
       byte_capacity_(byte_capacity),
-      id_(next_backing_store_id_.fetch_add(1)) {
+      id_(next_backing_store_id_.fetch_add(1)),
+      is_nodejs_(false) {
   // TODO(v8:11111): RAB / GSAB - Wasm integration.
   DCHECK_IMPLIES(is_wasm_memory64, is_wasm_memory);
   DCHECK_IMPLIES(has_guard_regions, is_wasm_memory);
@@ -210,6 +211,16 @@
     return;
   }
 
+  if (is_nodejs_) {
+    // JSArrayBuffer backing store. Deallocate through the embedder's allocator.
+    auto allocator = reinterpret_cast<v8::ArrayBuffer::Allocator*>(
+        get_v8_api_array_buffer_allocator());
+    TRACE_BS("BSn:free   bs=%p mem=%p (length=%zu, capacity=%zu)\n", this,
+             buffer_start_, byte_length(), byte_capacity_);
+    allocator->Free(buffer_start_, byte_length_, v8::ArrayBuffer::Allocator::AllocationMode::kNodeJS);
+    return;
+  }
+
   // JSArrayBuffer backing store. Deallocate through the embedder's allocator.
   auto allocator = get_v8_api_array_buffer_allocator();
   TRACE_BS("BS:free   bs=%p mem=%p (length=%zu, capacity=%zu)\n", this,
@@ -743,7 +754,7 @@
 std::unique_ptr<BackingStore> BackingStore::WrapAllocation(
     void* allocation_base, size_t allocation_length,
     v8::BackingStore::DeleterCallback deleter, void* deleter_data,
-    SharedFlag shared) {
+    SharedFlag shared, bool is_nodejs) {
   bool is_empty_deleter = (deleter == v8::BackingStore::EmptyDeleter);
   auto result = new BackingStore(allocation_base,               // start
                                  allocation_length,             // length
@@ -757,6 +768,7 @@
                                  false,              // has_guard_regions
                                  true,               // custom_deleter
                                  is_empty_deleter);  // empty_deleter
+  result->is_nodejs_ = is_nodejs;
   result->type_specific_data_.deleter = {deleter, deleter_data};
   TRACE_BS("BS:wrap   bs=%p mem=%p (length=%zu)\n", result,
            result->buffer_start(), result->byte_length());
diff -r -u --color up/v8/src/objects/backing-store.h nw/v8/src/objects/backing-store.h
--- up/v8/src/objects/backing-store.h	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/src/objects/backing-store.h	2026-04-25 12:56:39.700044916 +0000
@@ -80,7 +80,7 @@
   static std::unique_ptr<BackingStore> WrapAllocation(
       void* allocation_base, size_t allocation_length,
       v8::BackingStore::DeleterCallback deleter, void* deleter_data,
-      SharedFlag shared);
+      SharedFlag shared, bool is_nodejs = false);
 
   // Create an empty backing store.
   static std::unique_ptr<BackingStore> EmptyBackingStore(SharedFlag shared);
@@ -104,6 +104,7 @@
   bool is_immutable() const { return has_flag(kIsImmutable); }
   bool is_wasm_memory() const { return has_flag(kIsWasmMemory); }
   bool is_wasm_memory64() const { return has_flag(kIsWasmMemory64); }
+  bool is_node_js() const { return is_nodejs_; }
   bool has_guard_regions() const { return has_flag(kHasGuardRegions); }
 
   void set_is_immutable(bool immutable) {
@@ -124,6 +125,8 @@
   ResizeOrGrowResult ResizeInPlace(Isolate* isolate, size_t new_byte_length);
   ResizeOrGrowResult GrowInPlace(Isolate* isolate, size_t new_byte_length);
 
+  void set_nodejs(bool nodejs) { is_nodejs_ = nodejs; }
+
 #if V8_ENABLE_WEBASSEMBLY
   // The IsResizableByJs flag is set for backing stores for a resizable
   // ArrayBuffer or a WebAssembly.Memory that exposes its buffer as resizable
@@ -311,6 +314,7 @@
   } type_specific_data_;
 
   std::atomic<base::EnumSet<Flag, uint16_t>> flags_;
+  bool is_nodejs_ : 1;
 };
 
 // A global, per-process mapping from buffer addresses to backing stores
diff -r -u --color up/v8/src/objects/js-array-buffer-inl.h nw/v8/src/objects/js-array-buffer-inl.h
--- up/v8/src/objects/js-array-buffer-inl.h	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/src/objects/js-array-buffer-inl.h	2026-04-25 12:56:39.706624751 +0000
@@ -229,6 +229,8 @@
                     JSArrayBuffer::IsResizableByJsBit)
 BIT_FIELD_ACCESSORS(JSArrayBuffer, bit_field, is_immutable,
                     JSArrayBuffer::IsImmutableBit)
+BIT_FIELD_ACCESSORS(JSArrayBuffer, bit_field, is_node_js,
+                    JSArrayBuffer::IsNodejsBit)
 
 bool JSArrayBuffer::IsEmpty() const {
   auto backing_store = GetBackingStore();
diff -r -u --color up/v8/src/objects/js-array-buffer.h nw/v8/src/objects/js-array-buffer.h
--- up/v8/src/objects/js-array-buffer.h	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/src/objects/js-array-buffer.h	2026-04-25 12:56:39.706624751 +0000
@@ -74,6 +74,7 @@
 
   // [was_detached]: true => the buffer was previously detached.
   DECL_BOOLEAN_ACCESSORS(was_detached)
+  DECL_BOOLEAN_ACCESSORS(is_node_js)
 
   // [is_shared]: true if this is a SharedArrayBuffer or a
   // GrowableSharedArrayBuffer.
diff -r -u --color up/v8/src/objects/js-array-buffer.tq nw/v8/src/objects/js-array-buffer.tq
--- up/v8/src/objects/js-array-buffer.tq	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/src/objects/js-array-buffer.tq	2026-04-25 12:56:39.706624751 +0000
@@ -10,6 +10,7 @@
   is_shared: bool: 1 bit;
   is_resizable_by_js: bool: 1 bit;
   is_immutable: bool: 1 bit;
+  is_nodejs: bool: 1 bit;
 }
 
 const kNoView: Smi = 0;
diff -r -u --color up/v8/src/objects/js-function.cc nw/v8/src/objects/js-function.cc
--- up/v8/src/objects/js-function.cc	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/src/objects/js-function.cc	2026-04-25 12:56:39.709624723 +0000
@@ -1432,6 +1432,13 @@
     return NativeCodeFunctionSourceString(isolate, shared_info);
   }
 
+  //NWJS#6061: moved here or it will crash when trying to print
+  //function as a class
+  // Check if we have source code for the {function}.
+  if (!shared_info->HasSourceCode()) {
+    return NativeCodeFunctionSourceString(isolate, shared_info);
+  }
+
   if (IsClassConstructor(shared_info->kind())) {
     // Check if we should print {function} as a class.
     DirectHandle<Object> maybe_class_positions = JSReceiver::GetDataProperty(
@@ -1448,11 +1455,6 @@
     }
   }
 
-  // Check if we have source code for the {function}.
-  if (!shared_info->HasSourceCode()) {
-    return NativeCodeFunctionSourceString(isolate, shared_info);
-  }
-
   // If this function was compiled from asm.js, use the recorded offset
   // information.
 #if V8_ENABLE_WEBASSEMBLY
diff -r -u --color up/v8/src/objects/script.h nw/v8/src/objects/script.h
--- up/v8/src/objects/script.h	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/src/objects/script.h	2026-04-25 12:56:39.721624613 +0000
@@ -287,8 +287,10 @@
   friend FactoryBase<Factory>;
   friend FactoryBase<LocalFactory>;
 
+public:
   // Hide torque-generated accessor, use Script::SetSource instead.
   using TorqueGeneratedScript::set_source;
+private:
 
   // Bit positions in the flags field.
   DEFINE_TORQUE_GENERATED_SCRIPT_FLAGS()
diff -r -u --color up/v8/src/profiler/tracing-cpu-profiler.cc nw/v8/src/profiler/tracing-cpu-profiler.cc
--- up/v8/src/profiler/tracing-cpu-profiler.cc	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/src/profiler/tracing-cpu-profiler.cc	2026-04-25 12:56:39.732624513 +0000
@@ -28,6 +28,7 @@
 #if defined(V8_USE_PERFETTO)
   TrackEvent::RemoveSessionObserver(this);
 #else
+  if (V8::GetCurrentPlatform())
   V8::GetCurrentPlatform()->GetTracingController()->RemoveTraceStateObserver(
       this);
 #endif
@@ -79,9 +80,11 @@
   // It could be a long time until the Isolate next runs any JS which could be
   // interrupted, and we'd rather not leave the sampler thread running during
   // that time, so also post a task to run any interrupts.
+  std::shared_ptr<v8::TaskRunner> task_runner =
   V8::GetCurrentPlatform()
-      ->GetForegroundTaskRunner(reinterpret_cast<v8::Isolate*>(isolate_))
-      ->PostTask(std::make_unique<RunInterruptsTask>(isolate_));
+      ->GetForegroundTaskRunner(reinterpret_cast<v8::Isolate*>(isolate_));
+  if (task_runner) //NWJS#8266
+      task_runner->PostTask(std::make_unique<RunInterruptsTask>(isolate_));
 }
 
 void TracingCpuProfilerImpl::StartProfiling() {
diff -r -u --color up/v8/src/snapshot/code-serializer.cc nw/v8/src/snapshot/code-serializer.cc
--- up/v8/src/snapshot/code-serializer.cc	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/src/snapshot/code-serializer.cc	2026-04-25 12:56:39.748142316 +0000
@@ -530,6 +530,7 @@
   if (!maybe_result.ToHandle(&result)) {
     // Deserializing may fail if the reservations cannot be fulfilled.
     if (v8_flags.profile_deserialization) PrintF("[Deserializing failed]\n");
+    cached_data->Reject();
     return MaybeDirectHandle<SharedFunctionInfo>();
   }
 
@@ -862,9 +863,12 @@
     SerializedCodeSanityCheckResult* rejection_result) {
   DisallowGarbageCollection no_gc;
   SerializedCodeData scd(cached_data);
+  *rejection_result = SerializedCodeSanityCheckResult::kSuccess;
+#if 0
   *rejection_result = scd.SanityCheck(
       Snapshot::ExtractReadOnlySnapshotChecksum(isolate->snapshot_blob()),
       expected_source_hash);
+#endif
   if (*rejection_result != SerializedCodeSanityCheckResult::kSuccess) {
     cached_data->Reject();
     return SerializedCodeData(nullptr, 0);
diff -r -u --color up/v8/src/snapshot/deserializer.cc nw/v8/src/snapshot/deserializer.cc
--- up/v8/src/snapshot/deserializer.cc	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/src/snapshot/deserializer.cc	2026-04-25 12:56:39.748142316 +0000
@@ -352,7 +352,9 @@
 #ifdef DEBUG
   num_api_references_ = GetNumApiReferences(isolate);
 #endif  // DEBUG
-  CHECK_EQ(magic_number_, SerializedData::kMagicNumber);
+  bool ret =
+    (magic_number_ == SerializedData::kMagicNumber);
+  valid_ = ret;
 }
 
 template <typename IsolateT>
diff -r -u --color up/v8/src/snapshot/deserializer.h nw/v8/src/snapshot/deserializer.h
--- up/v8/src/snapshot/deserializer.h	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/src/snapshot/deserializer.h	2026-04-25 12:56:39.748142316 +0000
@@ -123,6 +123,8 @@
 
   DirectHandle<HeapObject> ReadObject();
 
+  bool valid_ = true;
+
  private:
   // A circular queue of hot objects. This is added to in the same order as in
   // Serializer::HotObjectsList, but this stores the objects as a vector of
diff -r -u --color up/v8/src/snapshot/object-deserializer.cc nw/v8/src/snapshot/object-deserializer.cc
--- up/v8/src/snapshot/object-deserializer.cc	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/src/snapshot/object-deserializer.cc	2026-04-25 12:56:39.748887923 +0000
@@ -36,6 +36,8 @@
 
 MaybeDirectHandle<HeapObject> ObjectDeserializer::Deserialize() {
   DCHECK(deserializing_user_code());
+  if (!valid_)
+    return MaybeHandle<HeapObject>();
   HandleScope scope(isolate());
   DirectHandle<HeapObject> result;
   {
diff -r -u --color up/v8/src/strings/string-hasher-inl.h nw/v8/src/strings/string-hasher-inl.h
--- up/v8/src/strings/string-hasher-inl.h	2026-04-17 08:21:16.000000000 +0000
+++ nw/v8/src/strings/string-hasher-inl.h	2026-04-25 12:56:39.751504077 +0000
@@ -60,7 +60,7 @@
   // For 2-byte strings we need to preserve the same hash for strings in just
   // the latin-1 range.
   if (V8_UNLIKELY(IsOnly8Bit(chars, length))) {
-    return detail::HashConvertingTo8Bit(chars, length, seed, secret);
+    return v8::internal::detail::HashConvertingTo8Bit(chars, length, seed, secret);
   }
   return rapidhash(reinterpret_cast<const uint8_t*>(chars), 2 * length, seed,
                    secret);
@@ -84,7 +84,7 @@
   running_hash_ += (running_hash_ << 3);
   running_hash_ ^= (running_hash_ >> 11);
   running_hash_ += (running_hash_ << 15);
-  return detail::ConvertRawHashToUsableHash(running_hash_);
+  return v8::internal::detail::ConvertRawHashToUsableHash(running_hash_);
 }
 
 uint32_t StringHasher::GetTrivialHash(uint32_t length) {
@@ -297,28 +297,28 @@
       // Possible array or integer index; try to compute the array index hash.
       static_assert(String::kMaxArrayIndexSize <= String::kMaxIntegerIndexSize);
 
-      detail::ArrayIndexT index;
+      v8::internal::detail::ArrayIndexT index;
       uint32_t i;
-      switch (detail::TryParseArrayIndex(chars, length, i, index)) {
-        case detail::kSuccess: {
+      switch (v8::internal::detail::TryParseArrayIndex(chars, length, i, index)) {
+        case v8::internal::detail::kSuccess: {
           DCHECK_LE(index, String::kMaxArrayIndex);
           return StringHasher::MakeArrayIndexHash(static_cast<uint32_t>(index),
                                                   length, seed);
         }
-        case detail::kNonIndex:
+        case v8::internal::detail::kNonIndex:
           // A non-index result from TryParseArrayIndex means we don't need to
           // check for integer indices.
           break;
-        case detail::kOverflow: {
+	      case v8::internal::detail::kOverflow: {
 #if V8_HOST_ARCH_64_BIT
           // On 64-bit, we might have a valid integer index even if the value
           // overflowed an array index.
           static_assert(String::kMaxArrayIndexSize <
                         String::kMaxIntegerIndexSize);
-          switch (detail::TryParseIntegerIndex(chars, length, i, index)) {
-            case detail::kSuccess: {
+          switch (v8::internal::detail::TryParseIntegerIndex(chars, length, i, index)) {
+		  case v8::internal::detail::kSuccess: {
               uint32_t hash = String::CreateHashFieldValue(
-                  detail::GetUsableRapidHash(chars, length, seed.seed(),
+                  v8::internal::detail::GetUsableRapidHash(chars, length, seed.seed(),
                                              seed.secret()),
                   String::HashFieldType::kIntegerIndex);
               if (Name::ContainsCachedArrayIndex(hash)) {
@@ -331,8 +331,8 @@
               DCHECK(!Name::ContainsCachedArrayIndex(hash));
               return hash;
             }
-            case detail::kNonIndex:
-            case detail::kOverflow:
+	    case v8::internal::detail::kNonIndex:
+	    case v8::internal::detail::kOverflow:
               break;
           }
 #else
@@ -354,7 +354,7 @@
 
   // Non-index hash.
   return String::CreateHashFieldValue(
-      detail::GetUsableRapidHash(chars, length, seed.seed(), seed.secret()),
+      v8::internal::detail::GetUsableRapidHash(chars, length, seed.seed(), seed.secret()),
       String::HashFieldType::kHash);
 }
 
Only in nw/v8/src: v8.gyp
